widget/list_test.go

```diff
@@ -0,0 +1,161 @@
+	template := fyne.NewContainerWithLayout(layout.NewHBoxLayout(), NewIcon(theme.DocumentIcon()), NewLabel("Template Object"))
+	test.WidgetRenderer(list).(*listRenderer).Layout(fyne.NewSize(100, 1000))
+
+	firstItemIndex := test.WidgetRenderer(list).(*listRenderer).firstItemIndex
+	lastItemIndex := test.WidgetRenderer(list).(*listRenderer).lastItemIndex
+	visibleCount := len(test.WidgetRenderer(list).(*listRenderer).children)
+	poolCount := test.WidgetRenderer(list).(*listRenderer).itemPool.Count()
+
+	assert.Equal(t, 0, firstItemIndex)
+	assert.Equal(t, visibleCount, lastItemIndex-firstItemIndex+1)
+	assert.Equal(t, poolCount, 5)
+
+	scroll := test.WidgetRenderer(list).(*listRenderer).scroller
+	scroll.Scrolled(&fyne.ScrollEvent{DeltaX: 0, DeltaY: -300})
+
+	indexChange := int(math.Floor(float64(300) / float64(template.MinSize().Height)))
+
+	newFirstItemIndex := test.WidgetRenderer(list).(*listRenderer).firstItemIndex
+	newLastItemIndex := test.WidgetRenderer(list).(*listRenderer).lastItemIndex
+	newVisibleCount := len(test.WidgetRenderer(list).(*listRenderer).children)
+	poolCount = test.WidgetRenderer(list).(*listRenderer).itemPool.Count()
+
+	assert.NotEqual(t, firstItemIndex, newFirstItemIndex)
+	assert.Equal(t, newFirstItemIndex, firstItemIndex+indexChange)
+	assert.NotEqual(t, lastItemIndex, newLastItemIndex)
+	assert.Equal(t, newLastItemIndex, lastItemIndex+indexChange)
+	assert.Equal(t, visibleCount, newVisibleCount)
+	assert.Equal(t, newVisibleCount, newLastItemIndex-newFirstItemIndex+1)
+	assert.Equal(t, poolCount, 5)
+}
+
+func TestListHover(t *testing.T) {
+	list := createList()
+	test.WidgetRenderer(list).(*listRenderer).Layout(fyne.NewSize(100, 1000))
+	children := test.WidgetRenderer(list).(*listRenderer).children
+
+	for i := 0; i < 2; i++ {
+		assert.Equal(t, children[i].(*listItem).statusIndicator.FillColor, theme.BackgroundColor())
+		children[i].(*listItem).MouseIn(&desktop.MouseEvent{})
+		assert.Equal(t, children[i].(*listItem).statusIndicator.FillColor, theme.HoverColor())
+		children[i].(*listItem).MouseOut()
+		assert.Equal(t, children[i].(*listItem).statusIndicator.FillColor, theme.BackgroundColor())
+	}
+}
+
+func TestListSelection(t *testing.T) {
+	list := createList()
+	test.WidgetRenderer(list).(*listRenderer).Layout(fyne.NewSize(100, 1000))
+	children := test.WidgetRenderer(list).(*listRenderer).children
+
+	assert.Equal(t, children[0].(*listItem).statusIndicator.FillColor, theme.BackgroundColor())
+	children[0].(*listItem).Tapped(&fyne.PointEvent{})
+	assert.Equal(t, children[0].(*listItem).statusIndicator.FillColor, theme.FocusColor())
+	assert.Equal(t, list.selectedIndex, 0)
+	children[1].(*listItem).Tapped(&fyne.PointEvent{})
+	assert.Equal(t, children[1].(*listItem).statusIndicator.FillColor, theme.FocusColor())
+	assert.Equal(t, list.selectedIndex, 1)
+	assert.Equal(t, children[0].(*listItem).statusIndicator.FillColor, theme.BackgroundColor())
+}
+
+func testListDataChange(t *testing.T) {
```

Comment:
Where is this used? It look identical to the function above. If this is a test it should be exported.

Reply:
This should be resolved now